node/api/wallet.go

```diff
@@ -179,9 +191,24 @@ func (api *API) walletAddressHandler(w http.ResponseWriter, req *http.Request, _
 
 // walletAddressHandler handles API calls to /wallet/addresses.
 func (api *API) walletAddressesHandler(w http.ResponseWriter, req *http.Request, _ httprouter.Params) {
-	WriteJSON(w, WalletAddressesGET{
-		Addresses: api.wallet.AllAddresses(),
-	})
+	paginationRequest := GetPaginationDefaultRequest(req)
+	allAddresses := api.wallet.AllAddresses()
+	if !paginationRequest.HasPaginationQuery || allAddresses == nil {
+		WriteJSON(w, WalletAddressesGET{
+			Addresses: allAddresses,
+		})
+	} else {
+		if !paginationRequest.PaginationQueryIsValid {
+			WriteError(w, Error{"start and limit queries must be nonnegative integers"}, http.StatusBadRequest)
+			return
+		}
+		pagination, paginationResponse := GetPaginationIndicesAndResponse(allAddresses, paginationRequest)
+		addresses := allAddresses[pagination.Start:pagination.End]
+		WriteJSON(w, WalletAddressesPaginatedGET{
+			Addresses:  addresses,
+			Pagination: paginationResponse,
+		})
+	}
```

Comment:
As mentioned above I would change the call to `GetPaginationRequest` so that you get something like the following:
```
start, limit, err := GetPaginationDefaultRequest(req)
...
if err == errNoPagination {
  // legacy behavior
  return
}
// new behavior

Reply:
You can create a custom error like `errNoPagination` by declaring it at the top of the file. You can find an example in `modules/transactionpool/transactionpool.go` where we declare `errNilCS` and `errNilGateway` at the top.
Of course that also means you have to do a `return errNoPagination` in `GetPaginationRequest` if no params were specified.